File: solutions/java/need-for-speed/1/src/main/java/NeedForSpeed.java
class NeedForSpeed {

    public int distance = 0;
    private int batteryRemaining = 100;
    public int batteryDrain;
    public int speed;
    
    
    NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() {
        return batteryDrain > batteryRemaining;
    }

    public int distanceDriven() {
        return distance;
    }

    public void drive() {
        if (batteryRemaining > 0){
            distance += speed;
            batteryRemaining -= batteryDrain;
        }
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {

    private int distance;
    
    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean canFinishRace(NeedForSpeed car) {
        return 100 / car.batteryDrain *  car.speed >= distance;
    }
}

File: solutions/csharp/need-for-speed/1/NeedForSpeed.cs
class RemoteControlCar
{
    // TODO: define the constructor for the 'RemoteControlCar' class

    public int speed;
    public int batteryDrain;
    private int distance = 0;
    private int battery = 100;

    public RemoteControlCar(int speed, int batteryDrain)
    {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public bool BatteryDrained() =>  battery < batteryDrain;

    public int DistanceDriven() =>  distance;


    public void Drive()
    {
        if (BatteryDrained())
        {
            return;
        }
        distance += speed;
        battery -= batteryDrain;
    }

    public static RemoteControlCar Nitro() =>  new RemoteControlCar(50, 4);

}

class RaceTrack
{
    // TODO: define the constructor for the 'RaceTrack' class

    private int distance;
    public RaceTrack(int distance)
    {
        this.distance = distance;
    }
    public bool TryFinishTrack(RemoteControlCar car)
    {
        int maxDrives = 100 / car.batteryDrain;
        int maxDistance = maxDrives * car.speed;

        return maxDistance >= distance;  
    }
}
